### PR TITLE
ETF MILESTONE 2

### PR DESCRIPTION
**Pull request by driemworks**

### CODE REVIEW

File: polkadot/runtime/rococo/src/lib.rs
Comment (juangirini) [inline polkadot/runtime/rococo/src/lib.rs:1]: I don't think we need to change Rococo, why is this needed?

File: polkadot/runtime/westend/src/lib.rs
Comment (juangirini) [inline polkadot/runtime/westend/src/lib.rs:1]: Same for Westend, why do we need to change it?

File: substrate/primitives/core/src/crypto_bytes.rs [hunk 1]
```diff
@@ -3,7 +3,7 @@
 // Copyright (C) Parity Technologies (UK) Ltd.
 // SPDX-License-Identifier: Apache-2.0
 
-// Licensed under the Apache License, Version 2.0 (the "License");
+// Licensed under the Apache License, 1sion 2.0 (the "License");
```
Comment (juangirini) [inline substrate/primitives/core/src/crypto_bytes.rs]:
```suggestion
// Licensed under the Apache License, Version 2.0 (the "License");
```

File: polkadot/runtime/rococo/src/lib.rs
Comment (driemworks) [reply polkadot/runtime/rococo/src/lib.rs:1]: In order to make resharings and public commitments available to BEEFY authorities, I modified the BEEFY api to be able to fetch this from the runtime. We can't *not* implement it in these files without getting compilation errors. I was planning on hiding it behind an "etf" feature after I get everything working, but we can probably remove these items as well as long as we only care about compiling the substrate package.

File: polkadot/runtime/westend/src/lib.rs
Comment (driemworks) [reply polkadot/runtime/westend/src/lib.rs:1]: same as above

File: substrate/bin/node/cli/src/chain_spec.rs [hunk 2]
```diff
@@ -353,7 +383,54 @@ fn configure_accounts(
...
+			).unwrap()
+		}).collect::<Vec<_>>(), 
+		initial_authorities.len() as u8, 
+		&mut OsRng,
+	);
+
+	initial_authorities.iter().enumerate().map(|(idx, auth)| {
+		let pok = &genesis_resharing[idx].1;
+		let mut bytes = Vec::new();
+		pok.serialize_compressed(&mut bytes).unwrap();
```
Comment (juangirini) [inline substrate/bin/node/cli/src/chain_spec.rs:426]: Do we need to unwrap this?

File: substrate/client/consensus/beefy/src/communication/gossip.rs [hunk 3]
```diff
@@ -303,10 +304,30 @@ where
 				return Action::Discard(cost::UNKNOWN_VOTER)
 			}
 		}
-
-		if BeefyKeystore::verify(&vote.id, &vote.signature, &vote.commitment.encode()) {
+	
+		// recover the signature bytes from the payload
+		let raw_etf_payload = vote.commitment.payload.get_raw(
+				&sp_consensus_beefy::known_payloads::ETF_SIGNATURE
+			).expect("its ok for now");
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/communication/gossip.rs]: ??? 😄

File: substrate/client/consensus/beefy/src/keystore.rs [hunk 4]
```diff
@@ -404,7 +485,7 @@ pub mod tests {
 
 		let store: BeefyKeystore<AuthorityId> = Some(store).into();
 
-		let msg = b"are you involved or committed?";
+		let msg = b"are you involved or commited?";
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/keystore.rs]: I'm pretty sure it is "committed"

File: substrate/client/consensus/beefy/src/lib.rs [hunk 5]
```diff
@@ -342,6 +343,7 @@ where
 		let mut sessions = VecDeque::new();
 		let mut header = best_grandpa.clone();
 		let state = loop {
+
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/lib.rs]: This is just a nit, but in general, we should avoid unnecessary diff changes
```suggestion
```

File: substrate/client/consensus/beefy/src/worker.rs [hunk 6]
```diff
@@ -76,13 +83,13 @@ pub(crate) enum RoundAction {
 pub(crate) struct VoterOracle<B: Block> {
 	/// Queue of known sessions. Keeps track of voting rounds (block numbers) within each session.
 	///
-	/// There are three voter states corresponding to three queue states:
+	/// There are three voter states coresponding to three queue states:
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/worker.rs]: double r is correct

File: substrate/frame/beefy/src/lib.rs [hunk 7]
```diff
@@ -444,6 +419,52 @@ impl<T: Config> Pallet<T> {
 
 		Ok(())
 	}
+
+	fn initialize_genesis_shares(
```
Comment (juangirini) [inline substrate/frame/beefy/src/lib.rs:484]: I guess it is fine for this fn to panic?

File: substrate/bin/node/cli/src/chain_spec.rs [hunk 8]
```diff
@@ -353,7 +383,55 @@ fn configure_accounts(
...
+	let mut ibe_pp_bytes = Vec::new();
+	ibe_pub_param.serialize_compressed(&mut ibe_pp_bytes).unwrap();
+
+	let genesis_resharing = double_secret.reshare(
+		&initial_authorities.iter().map(|authority| {
+			w3f_bls::single::PublicKey::<EB>::from_bytes(&authority.to_raw_vec()[..48]).unwrap()
+			// EB::SignatureGroup::deserialize_compressed(
+			// 	// [48 bytes for SigGroup][96 bytes for PubKeyGroup]
+			// 	&authority.to_raw_vec()[..48]
+			// ).unwrap()
```
Comment (juangirini) [inline substrate/bin/node/cli/src/chain_spec.rs]:
```suggestion
```

File: substrate/bin/node/cli/src/chain_spec.rs [hunk 9]
```diff
@@ -438,7 +522,11 @@ pub fn testnet_genesis(
 
 fn development_config_genesis_json() -> serde_json::Value {
 	testnet_genesis(
-		vec![authority_keys_from_seed("Alice")],
+		vec![
+			authority_keys_from_seed("Alice"), 
+			// authority_keys_from_seed("Bob"), 
+			// authority_keys_from_seed("Charlie")
```
Comment (juangirini) [inline substrate/bin/node/cli/src/chain_spec.rs:530]:
```suggestion
```

File: substrate/client/consensus/beefy/Cargo.toml [hunk 10]
```diff
@@ -37,10 +37,14 @@ sp-consensus = { path = "../../../primitives/consensus/common" }
 sp-consensus-beefy = { path = "../../../primitives/consensus/beefy" }
 sp-core = { path = "../../../primitives/core" }
 sp-crypto-hashing = { path = "../../../primitives/crypto/hashing" }
-sp-keystore = { path = "../../../primitives/keystore" }
+sp-keystore = { path = "../../../primitives/keystore", features = ["bls-experimental"]}
 sp-mmr-primitives = { path = "../../../primitives/merkle-mountain-range" }
 sp-runtime = { path = "../../../primitives/runtime" }
 tokio = "1.22.0"
+# etf-crypto-primitives = { git = "https://github.com/ideal-lab5/etf-sdk.git", branch = "w3fbls-migration" }
```
Comment (juangirini) [inline substrate/client/consensus/beefy/Cargo.toml]: cleanup

File: substrate/bin/node/cli/Cargo.toml [hunk 11]
```diff
@@ -118,6 +118,15 @@ node-inspect = { package = "staging-node-inspect", path = "../inspect", optional
 try-runtime-cli = { path = "../../../utils/frame/try-runtime/cli", optional = true }
 serde_json = { workspace = true, default-features = true }
 
+# etf dependencies
+etf-crypto-primitives = { git = "https://github.com/ideal-lab5/etf-sdk.git", branch = "w3fbls-migration" }
```
Comment (juangirini) [inline substrate/bin/node/cli/Cargo.toml:122]: nice :)

File: substrate/client/consensus/beefy/src/communication/gossip.rs [hunk 12]
```diff
@@ -303,8 +304,34 @@ where
 				return Action::Discard(cost::UNKNOWN_VOTER)
 			}
 		}
+	
+		// recover the signature bytes from the payload
+		let raw_etf_payload = vote.commitment.payload.get_raw(&sp_consensus_beefy::known_payloads::ETF_SIGNATURE)
+			.unwrap_or({
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/communication/gossip.rs]: noice!

File: substrate/client/consensus/beefy/src/communication/gossip.rs [hunk 13]
```diff
@@ -303,8 +304,34 @@ where
...
+					debug!(
+						target: LOG_TARGET,
+						"🎲 Corrupted (irrecoverable) signature on message: {:?}, from: {:?}", vote, sender
+					);
+					return Action::Discard(cost::BAD_SIGNATURE);
+				});
 
-		if BeefyKeystore::verify(&vote.id, &vote.signature, &vote.commitment.encode()) {
+		if BeefyKeystore::verify(&vote.id, &vote.signature, &vote.commitment.encode())
+			// && BeefyKeystore::verify(&vote.id, &etf_sig, &round.to_string().as_bytes())
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/communication/gossip.rs]:
```suggestion
```

File: substrate/client/consensus/beefy/src/keystore.rs [hunk 14]
```diff
@@ -100,6 +107,18 @@ impl<AuthorityId: AuthorityIdBound> BeefyKeystore<AuthorityId> {
 			},
 
 			#[cfg(feature = "bls-experimental")]
+			bls377::CRYPTO_ID => {
+				let public: bls377::Public =
+					bls377::Public::try_from(public.as_slice()).unwrap();
+				let sig = store
+					.bls377_sign(BEEFY_KEY_TYPE, &public, &message)
+					.map_err(|e| error::Error::Keystore(e.to_string()))?
+					.ok_or_else(|| error::Error::Signature("bls377_sign()  failed".to_string()))?;
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/keystore.rs]:
```suggestion
					.ok_or_else(|| error::Error::Signature("bls377_sign() failed".to_string()))?;
```

File: substrate/client/consensus/beefy/src/worker.rs [hunk 15]
```diff
@@ -244,23 +251,26 @@ impl<B: Block> VoterOracle<B> {
...
+		// let target =
+		// 	vote_target(best_grandpa, best_beefy, rounds.session_start(), self.min_block_delta);
+		// info!(
+		// 	target: LOG_TARGET,
+		// 	"🥩 best beefy: #{:?}, best finalized: #{:?}, current_vote_target: {:?}",
+		// 	best_beefy,
+		// 	best_grandpa,
+		// 	target
+		// );
+		// target
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/worker.rs]: cleanup
```suggestion
```

File: substrate/client/consensus/beefy/src/worker.rs [hunk 16]
```diff
@@ -810,6 +830,35 @@ where
...
+			if let Some(Some(pok_bytes)) = runtime_api.read_share(hash, id.clone()).ok() {
+				if let Ok(sig)  = self.key_store.etf_sign(
+					&id, 
+					&pok_bytes, 
+					&message, validator_set.len() as u8
+				) {
+					return Some(sig);
+				}
+			}
+		}
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/worker.rs:923]: Optional: I usually find `match` statements easier to read than nested `if`s

```suggestion
		match (runtime_api.validator_set(hash), runtime_api.read_share(hash, id.clone()) {
			(Some(validator_set), Some(pok_bytes)) if let Ok(sig)  = self.key_store.etf_sign(
				&id, 
				&pok_bytes, 
				&message, validator_set.len() as u8
			) => return Some(sig);
			_ => return None;
		}
```

File: substrate/client/keystore/src/local.rs [hunk 17]
```diff
@@ -405,20 +406,38 @@ impl Keystore for LocalKeystore {
...
+			pok_bytes: &[u8],
+			message: &[u8],
+			threshold: u8,
+		) -> std::result::Result<bls377::Signature, TraitError>  {
+			if let Some(Some(etf_pair)) = self.0.read()
+				.key_pair_by_type::<bls377::Pair>(public, key_type)?
+				.map(|pair| pair.acss_recover(pok_bytes, threshold)) {
+				// "IBE.Extract" Q = s*H(message) + DLEQ Proof
+				let extract = etf_pair.sign(&message);
+				// let pk = etf_pair.public();
```
Comment (juangirini) [inline substrate/client/keystore/src/local.rs]: cleanup
```suggestion
```

File: substrate/primitives/consensus/beefy/Cargo.toml [hunk 18]
```diff
@@ -32,6 +32,8 @@ lazy_static = { version = "1.4.0", optional = true }
 [dev-dependencies]
 array-bytes = "6.1"
 w3f-bls = { version = "0.1.3", features = ["std"] }
+# w3f-bls = { path = "../../../../../bls", features = ["std"] }
+# w3f-bls = { git = "https://github.com/driemworks/bls.git", branch = "etf", features  = ["std"] }
```
Comment (juangirini) [inline substrate/primitives/consensus/beefy/Cargo.toml]:
```suggestion
```

File: substrate/primitives/consensus/beefy/src/lib.rs [hunk 19]
```diff
@@ -200,16 +200,23 @@ pub mod ecdsa_bls_crypto {
 		fn verify(&self, signature: &<Self as RuntimeAppPublic>::Signature, msg: &[u8]) -> bool {
 			// We can not simply call
 			// `EcdsaBlsPair::verify(signature.as_inner_ref(), msg, self.as_inner_ref())`
-			// because that invokes ECDSA default verification which performs Blake2b hash
+			// because that invokes ECDSA default verification which perfoms Blake2b hash
```
Comment (juangirini) [inline substrate/primitives/consensus/beefy/src/lib.rs]:
```suggestion
			// because that invokes ECDSA default verification which performs Blake2b hash
```

File: substrate/primitives/consensus/beefy/src/lib.rs [hunk 20]
```diff
@@ -200,16 +200,23 @@ pub mod ecdsa_bls_crypto {
...
+			false
+			// EcdsaBlsPair::verify::<H>(
+			// 	signature.as_inner_ref(),
+			// 	msg,
+			// )
+			// EcdsaBlsPair::verify_with_hasher::<H>(
+			// 	signature.as_inner_ref(),
+			// 	msg,
+			// 	self.as_inner_ref(),
+			// )
```
Comment (juangirini) [inline substrate/primitives/consensus/beefy/src/lib.rs]:
```suggestion
```

File: substrate/primitives/consensus/beefy/src/lib.rs [hunk 21]
```diff
@@ -538,7 +551,8 @@ mod tests {
 	}
 
 	#[test]
-	#[cfg(feature = "bls-experimental")]
+	// #[cfg(feature = "bls-experimental")]
```
Comment (juangirini) [inline substrate/primitives/consensus/beefy/src/lib.rs:554]:
```suggestion
```

File: substrate/primitives/core/Cargo.toml [hunk 22]
```diff
@@ -64,6 +64,11 @@ secp256k1 = { version = "0.28.0", default-features = false, features = ["alloc",
 w3f-bls = { version = "0.1.3", default-features = false, optional = true }
 # bandersnatch crypto
 bandersnatch_vrfs = { git = "https://github.com/w3f/ring-vrf", rev = "e9782f9", default-features = false, features = ["substrate-curves"], optional = true }
+# etf algs
+# etf-crypto-primitives = { path = "../../../../etf-sdk/etf-crypto-primitives", default-features = false, optional = true }
```
Comment (juangirini) [inline substrate/primitives/core/Cargo.toml]:
```suggestion
```

File: substrate/primitives/core/src/bls.rs [hunk 23]
```diff
@@ -414,4 +439,9 @@ mod tests {
 		// Poorly-sized
 		assert!(deserialize_signature("\"abc123\"").is_err());
 	}
+
+	#[test]
+	fn acss_recover_works() {
+
```
Comment (juangirini) [inline substrate/primitives/core/src/bls.rs]: I guess we need a test here? 😄

File: substrate/primitives/core/src/bls.rs [hunk 24]
```diff
@@ -157,7 +182,7 @@ impl<T: BlsBound> TraitPair for Pair<T> {
 
 	fn derive<Iter: Iterator<Item = DeriveJunction>>(
 		&self,
-		path: Iter,
+		path: Iter,
```
Comment (juangirini) [inline substrate/primitives/core/src/bls.rs]:
```suggestion
		path: Iter,
```

File: substrate/primitives/core/src/crypto_bytes.rs [hunk 25]
```diff
@@ -3,7 +3,7 @@
 // Copyright (C) Parity Technologies (UK) Ltd.
 // SPDX-License-Identifier: Apache-2.0
 
-// Licensed under the Apache License, Version 2.0 (the "License");
+// Licensed under the Apache License, version 2.0 (the "License");
```
Comment (juangirini) [inline substrate/primitives/core/src/crypto_bytes.rs]:
```suggestion
// Licensed under the Apache License, Version 2.0 (the "License");
```

File: substrate/frame/beefy-mmr/src/lib.rs [hunk 26]
```diff
@@ -76,6 +76,7 @@ where
 pub struct BeefyEcdsaToEthereum;
 impl Convert<sp_consensus_beefy::ecdsa_crypto::AuthorityId, Vec<u8>> for BeefyEcdsaToEthereum {
 	fn convert(beefy_id: sp_consensus_beefy::ecdsa_crypto::AuthorityId) -> Vec<u8> {
+		// Vec::new()
```
Comment (juangirini) [inline substrate/frame/beefy-mmr/src/lib.rs]:
```suggestion
```

File: substrate/frame/beefy-mmr/src/lib.rs [hunk 27]
```diff
@@ -86,6 +87,22 @@ impl Convert<sp_consensus_beefy::ecdsa_crypto::AuthorityId, Vec<u8>> for BeefyEc
...
+impl Convert<sp_consensus_beefy::bls_crypto::AuthorityId, Vec<u8>> for BeefyBlsToEthereum {
+	fn convert(_beefy_id: sp_consensus_beefy::bls_crypto::AuthorityId) -> Vec<u8> {
+		Vec::new()
+		// sp_core::ecdsa::Public::from(beefy_id)
+		// 	.to_eth_address()
+		// 	.map(|v| v.to_vec())
+		// 	.map_err(|_| {
+		// 		log::debug!(target: "runtime::beefy", "Failed to convert BEEFY PublicKey to ETH address!");
+		// 	})
+		// 	.unwrap_or_default()
```
Comment (juangirini) [inline substrate/frame/beefy-mmr/src/lib.rs:101]:
```suggestion
```

File: substrate/frame/beefy-mmr/src/lib.rs [hunk 28]
```diff
@@ -86,6 +87,22 @@ impl Convert<sp_consensus_beefy::ecdsa_crypto::AuthorityId, Vec<u8>> for BeefyEc
 	}
 }
 
+/// Convert BEEFY secp256k1 public keys into Ethereum addresses
+pub struct BeefyBlsToEthereum;
+impl Convert<sp_consensus_beefy::bls_crypto::AuthorityId, Vec<u8>> for BeefyBlsToEthereum {
+	fn convert(_beefy_id: sp_consensus_beefy::bls_crypto::AuthorityId) -> Vec<u8> {
+		Vec::new()
```
Comment (juangirini) [inline substrate/frame/beefy-mmr/src/lib.rs:94]: Do we need this? We aren't using it as far as I can see

File: substrate/frame/beefy/src/lib.rs [hunk 29]
```diff
@@ -55,10 +54,11 @@ mod tests;
 pub use crate::equivocation::{EquivocationOffence, EquivocationReportSystem, TimeSlot};
 pub use pallet::*;
 
-use crate::equivocation::EquivocationEvidenceFor;
+use crate::equivocation::{EquivocationEvidenceFor};
```
Comment (juangirini) [inline substrate/frame/beefy/src/lib.rs]:
```suggestion
use crate::equivocation::EquivocationEvidenceFor;
```

File: substrate/frame/beefy/src/lib.rs [hunk 30]
```diff
@@ -55,10 +54,11 @@ mod tests;
...
 
 const LOG_TARGET: &str = "runtime::beefy";
 
+
```
Comment (juangirini) [inline substrate/frame/beefy/src/lib.rs]:
```suggestion
```

File: substrate/frame/beefy/src/lib.rs [hunk 31]
```diff
@@ -133,6 +133,22 @@ pub mod pallet {
 	pub type NextAuthorities<T: Config> =
 		StorageValue<_, BoundedVec<T::BeefyId, T::MaxAuthorities>, ValueQuery>;
 
+	/// publicly verifiable shares for the current round (a resharing)
+	#[pallet::storage]
+	pub type Shares<T: Config> = 
+		StorageValue<_, BoundedVec<BoundedVec<u8, ConstU32<1024>>, T::MaxAuthorities>, ValueQuery>;
```
Comment (juangirini) [inline substrate/frame/beefy/src/lib.rs:139]: Is `BoundedVec<u8, ConstU32<1024>>` a share? Would it make sense to create a type for it?

File: substrate/frame/beefy/src/lib.rs [hunk 32]
```diff
@@ -444,6 +413,42 @@ impl<T: Config> Pallet<T> {
 
 		Ok(())
 	}
+
+	fn initialize_genesis_shares(
+		genesis_resharing: &Vec<(T::BeefyId, T::BeefyId, Vec<u8>)>,
+		round_key: Vec<u8>,
+	) -> Result<(), ()>  {
+		// TODO: define 96 as a const
```
Comment (juangirini) [inline substrate/frame/beefy/src/lib.rs:488]: and `BoundedVec<u8, ConstU32<1024>>` as a type?

File: substrate/bin/node/runtime/src/lib.rs [hunk 33]
```diff
@@ -3011,6 +3011,38 @@ impl_runtime_apis! {
...
+				return Some(commitments[at as usize].clone());
+			}
+			None
+
+
+			// let commitments = pallet_beefy::Commitments::<Runtime>::get();
+			// if at as usize >= commitments.len() {
+			// 	return None;
+			// }
+			// Some(commitments[at as usize].clone().into_inner())
```
Comment (juangirini) [inline substrate/bin/node/runtime/src/lib.rs]:
```suggestion
```

File: substrate/client/consensus/beefy/src/keystore.rs [hunk 34]
```diff
@@ -16,16 +16,23 @@
...
+	key_types::ETF as ETF_KEY_TYPE,
+	AppCrypto, 
+	RuntimeAppPublic
+};
 use sp_consensus_beefy::{AuthorityIdBound, BeefyAuthorityId, BeefySignatureHasher};
 use sp_core::ecdsa;
 #[cfg(feature = "bls-experimental")]
+use sp_core::bls377;
+#[cfg(feature = "bls-experimental")]
 use sp_core::ecdsa_bls377;
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/keystore.rs]:
```suggestion
#[cfg(feature = "bls-experimental")]
use sp_core::{bls377, ecdsa_bls377};
```

File: substrate/client/consensus/beefy/src/lib.rs [hunk 35]
```diff
@@ -397,6 +401,7 @@ where
 				.ok_or_else(|| Error::Backend("Invalid BEEFY chain".into()))?
 			}
 
+
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/lib.rs]:
```suggestion
```

File: substrate/client/consensus/beefy/src/worker.rs [hunk 36]
```diff
@@ -747,23 +753,38 @@ where
...
+				warn!(target: LOG_TARGET, "🥩 Error calculating ETF signature");
 				return Ok(())
 			},
 		};
+		
+		let etf_authority_id = self.runtime.runtime_api()
+			.read_commitment(target_hash, authority_id.clone())
+			.ok()
+			.unwrap()
+			.expect("todo");
```
Comment (juangirini) [inline substrate/client/consensus/beefy/src/worker.rs]: Would it make sense to return an Error here?

**Review by juangirini — COMMENTED** [review]
It looks good, just some small comments

File: substrate/bin/node/cli/src/chain_spec.rs (hunk 2 above)
Comment (driemworks) [reply substrate/bin/node/cli/src/chain_spec.rs:426]: We do. It could be an expect instead, but it's fine if this panics since it's a critical failure if the pok can't be serialized on genesis anyway.

**Comment by driemworks** [comment]
Outdated. 
We decided to not go this route.
